File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-4.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func mergeTwoLists(_ list1: ListNode?, _ list2: ListNode?) -> ListNode? {
        var cur1: ListNode? = list1
        var cur2: ListNode? = list2
        let dummy: ListNode? = ListNode(-1)
        var cur: ListNode? = dummy

       

            while let node1 = cur1, let  node2 = cur2{

                if node1.val <= node2.val{
                    cur?.next = node1
                    cur1 = node1.next
                } else{
                    cur?.next = node2
                    cur2 = node2.next
                }
                 cur = cur?.next
            }

            // if cur1!.val <= cur2!.val {

            //     cur?.next = cur1
            //     cur1 = cur1?.next
                
                
            // }else {
            //     cur?.next = cur2
                
            //     cur2 = cur2?.next
            // }

            // cur = cur?.next

        // }

        if(cur1 == nil){
            cur?.next = cur2
        }
        if(cur2 == nil){
            cur?.next = cur1
        }

        return dummy?.next

    }
}

File: Data Structures & Algorithms/reverse-a-linked-list/submission-4.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode cur= head, prev = null;

        // imp:- return prev
        // so u will keep shifting the next like to the prev by 2
        // pointers us eprev = null, temp, cur = head
        // swap them keep makimg prev as cur and cur as cur.next
        
        // while(cur!=null){
        //     temp = cur.next;
        //     cur.next = prev;
        //     prev = cur;
        //     cur = temp;
            
        // }
        // return prev;

        while(cur!=null){
            ListNode temp = cur.next;
            cur.next = prev;
            prev=cur;
            cur = temp;
        }

        return prev;
    }
}
